jobqueue_features/mpi_wrapper.py

```diff
@@ -92,3 +93,92 @@ def get_default_mpi_params(
             )
         )
     return {"cmd": cmd, "out": out, "err": err}
+
+
+def deserialize_and_execute(serialized_object):
+    # Ensure the serialized object is of the expected type
+    if isinstance(serialized_object, dict):
+        # Make sure it has the expected entries
+        if not ("header" in serialized_object and "frames" in serialized_object):
+            raise RuntimeError(
+                "serialized_object dict does not have expected keys [header, frames]"
+            )
+    else:
+        raise RuntimeError("Cannot deserialize without a serialized_object")
+    func = deserialize(serialized_object["header"], serialized_object["frames"])
+    if serialized_object.get("args_header"):
+        args = deserialize(
+            serialized_object["args_header"], serialized_object["args_frames"]
+        )
+    else:
+        args = False
+    if serialized_object.get("kwargs_header"):
+        kwargs = deserialize(
+            serialized_object["kwargs_header"], serialized_object["kwargs_frames"]
+        )
+    else:
+        kwargs = False
+
+    # Free memory space used by (potentially large) serialised object
+    del serialized_object
+
+    # Execute the function
+    if args and kwargs:
+        result = func(*args, **kwargs)
+    elif args:
+        result = func(*args)
+    elif kwargs:
+        result = func(**kwargs)
+    else:
+        result = func()
```

Comment:
Here, I would go for default values of args and kwargs if not provided, i.e. empty tuple for args and empty dictionary for kwargs. With that we can end up with very only `result = func(*args, **kwargs)`.

Reply:
Is that always valid? The specific problem I ran in to was that  `MPI.Finalize()` expects no arguments, is an empty list an argument? I guess I will try it and see.

Reply:
Checked it, it doesn't work:
```
Traceback (most recent call last):
  File "mpi_pickling.py", line 60, in <module>
    mpi_deserialize_and_execute()
  File "/p/home/jusers/ocais1/jureca/jobqueue_features/jobqueue_features/mpi_wrapper.py", line 180, in mpi_deserialize_and_execute
    result = deserialize_and_execute(serialized_object)
  File "/p/home/jusers/ocais1/jureca/jobqueue_features/jobqueue_features/mpi_wrapper.py", line 134, in deserialize_and_execute
    result = func()
  File "mpi_pickling.py", line 45, in shutdown
    MPI.Finalize([], {})
TypeError: Finalize() takes exactly 0 positional arguments (2 given)
```

Reply:
What about `*[], **{}`?

Reply:
If the python function takes no arguments, still the execution in `func(*[], **{})` should work. That was my intention in assigning to `args` `tuple()` or `[]` istead of `False`. Same for `kwargs` bu with dictionary.

Reply:
Yes, that works. Ok I will update.